File: 76-minimum-window-substring/minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    if (s.length < t.length) return ''

    const map = new Array(128).fill(0)
    let count = t.length
    let start = 0
    let minStart = 0
    let minLen = Infinity

    for (let c of t) {
        map[c.charCodeAt(0)]++
    }

    for (let end = 0; end < s.length; end++) {
        if (map[s.charCodeAt(end)]-- > 0) count--

        while (count === 0) {
            if (end - start + 1 < minLen) {
                minStart = start
                minLen = end - start + 1
            }

            if (map[s.charCodeAt(start++)]++ === 0) count++
        }
    }

    return minLen === Infinity ? "" : s.slice(minStart, minStart + minLen)
};

File: 76-minimum-window-substring/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(t) > len(s) or not s or not t:
            return ''

        t_counts = Counter(t)
        required = len(t_counts)

        left = 0
        formed = 0
        window_counts = defaultdict(int)

        min_len = float('inf')
        result_indices = (0, 0)

        for right, char in enumerate(s):
            window_counts[char] += 1

            if char in t_counts and window_counts[char] == t_counts[char]:
                formed += 1

            while left <= right and formed == required:
                current_len = right - left + 1

                if current_len < min_len:
                    min_len = current_len
                    result_indices = (left, right)

                left_char = s[left]
                window_counts[left_char] -= 1

                if left_char in t_counts and window_counts[left_char] < t_counts[left_char]:
                    formed -= 1
                
                left += 1

        if min_len == float('inf'):
            return ''
        else:
            start, end = result_indices
            return s[start : end + 1]

